### PR TITLE
Alibaba:  add metadata server IP validation

### PR DESCRIPTION
**Pull request by bd233**
Validate if  machine, service and cluster networking CIDR  contains metadata server IP

### CODE REVIEW

**Comment by bd233** [comment]
@patrickdillon I have updated here

File: pkg/types/alibabacloud/validation/platform.go [hunk 1]
```diff
@@ -1,20 +1,49 @@
...
+	allErrs := field.ErrorList{}
+	metadataServerIP := "100.100.100.200"
+	networkPath := field.NewPath("networking")
+	machineNetworkPath := networkPath.Child("machineNetwork")
+	clusterNetworkPath := networkPath.Child("clusterNetwork")
+	serviceNetworkPath := networkPath.Child("serviceNetwork")
+
+	for _, network := range n.MachineNetwork {
+		if network.CIDR.Contains(net.ParseIP(metadataServerIP)) {
+			allErrs = append(allErrs, field.Invalid(machineNetworkPath, metadataServerIP, fmt.Errorf("the contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```
Comment (patrickdillon) [inline pkg/types/alibabacloud/validation/platform.go:35]:
```suggestion
			allErrs = append(allErrs, field.Invalid(machineNetworkPath, metadataServerIP, fmt.Errorf("contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```

File: pkg/types/alibabacloud/validation/platform.go [hunk 2]
```diff
@@ -1,20 +1,49 @@
...
+		}
+	}
+	for _, network := range n.ClusterNetwork {
+		if network.CIDR.Contains(net.ParseIP(metadataServerIP)) {
+			allErrs = append(allErrs, field.Invalid(clusterNetworkPath, metadataServerIP, fmt.Errorf("the contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```
Comment (patrickdillon) [inline pkg/types/alibabacloud/validation/platform.go:40]:
```suggestion
			allErrs = append(allErrs, field.Invalid(clusterNetworkPath, metadataServerIP, fmt.Errorf("contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```

File: pkg/types/alibabacloud/validation/platform.go [hunk 3]
```diff
@@ -1,20 +1,49 @@
...
+		}
+	}
+	for _, network := range n.ServiceNetwork {
+		if network.Contains(net.ParseIP(metadataServerIP)) {
+			allErrs = append(allErrs, field.Invalid(serviceNetworkPath, metadataServerIP, fmt.Errorf("the contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```
Comment (patrickdillon) [inline pkg/types/alibabacloud/validation/platform.go:45]:
```suggestion
			allErrs = append(allErrs, field.Invalid(serviceNetworkPath, metadataServerIP, fmt.Errorf("contains %s which is reserved for the metadata service", metadataServerIP).Error()))
```

**Review by patrickdillon — COMMENTED** [review]
Typically we have tests for the install config and validation like this would have a test. You can use AWS as an example: https://github.com/openshift/installer/blob/master/pkg/types/aws/validation/platform_test.go

**Comment by bd233** [comment]
> Typically we have tests for the install config and validation like this would have a test. You can use AWS as an example: https://github.com/openshift/installer/blob/master/pkg/types/aws/validation/platform_test.go

I have updated

**Comment by kwoodson** [comment]
@patrickdillon Please TAL

**Pull request merged**